### PR TITLE
fix: [PL-38885]: create a service account with same name in a different scope

### PR DESCRIPTION
**Pull request by sahibaCCM**
## Describe your changes
unique_sa_idx will get dropped via Index manager in the second deployment; this behaviour is due to pre-existing test in index manger.
serviceAccountsPrimaryKey index will get created in the same deployement

Tested for:
-> created service account with same name on all scopes.

image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

https://github.com/harness/harness-core/assets/112424209/75c74fde-b2d3-436a-a686-2e404dc6d291

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>
<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48077)

### CODE REVIEW

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger uts

**Review by Jimit09Gandhi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48077)* status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java` line 70 at r4](https://reviewable.io/reviews/harness/harness-core/48077#-NW7TbYFEegwTFu9ZLb0:-NW7TbYFEegwTFu9ZLb1:b-oij8nn) ([raw file](https://github.com/harness/harness-core/blob/aa7d97acaa870f29e0ed97d9d9adef37615f46c0/120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java#L70)):*
> ```Java
>         .add(CompoundMongoIndex.builder()
>                  .name("unique_sa_idx")
>                  .field(ServiceAccountKeys.accountIdentifier)
> ```

@sahibaCCM   If we introducing new index is unique_sa_idx still needed?

File: 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java [hunk 1]
```diff
@@ -58,10 +58,17 @@ public class ServiceAccount implements PersistentEntity, UuidAware, NGAccountAcc
...
-                 .name("unique_sa_idx")
+                 .name("serviceAccountsPrimaryKey")
                  .field(ServiceAccountKeys.accountIdentifier)
                  .field(ServiceAccountKeys.identifier)
+                 .field(ServiceAccountKeys.orgIdentifier)
+                 .field(ServiceAccountKeys.projectIdentifier)
                  .unique(true)
+                 .build())
+        .add(CompoundMongoIndex.builder()
+                 .name("unique_sa_idx")
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java:69]: lets remove this one

File: 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java [hunk 2]
```diff
@@ -58,10 +58,17 @@ public class ServiceAccount implements PersistentEntity, UuidAware, NGAccountAcc
   public static List<MongoIndex> mongoIndexes() {
     return ImmutableList.<MongoIndex>builder()
         .add(CompoundMongoIndex.builder()
-                 .name("unique_sa_idx")
+                 .name("serviceAccountsPrimaryKey")
                  .field(ServiceAccountKeys.accountIdentifier)
                  .field(ServiceAccountKeys.identifier)
+                 .field(ServiceAccountKeys.orgIdentifier)
+                 .field(ServiceAccountKeys.projectIdentifier)
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java:65]: Lets have this index as 
                .field(ServiceAccountKeys.accountIdentifier)
                 .field(ServiceAccountKeys.orgIdentifier)
                 .field(ServiceAccountKeys.projectIdentifier)
                 .field(ServiceAccountKeys.identifier)

File: 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java (hunk 1 above)
Comment (sahibaCCM) [reply 120-ng-manager/src/main/java/io/harness/ng/serviceaccounts/entities/ServiceAccount.java:69]: SA is not being fetched on account level as a whole like we do for projects. On different level it has to be accessed individually.

**Comment by sahibaCCM** [comment]
trigger uts

**Review by meenakshi-harness — APPROVED** [review]

**Comment by sahibaCCM** [comment]
trigger uts

**Review by Jimit09Gandhi — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48077)* status: 2 file(s) to review. 1. unresolved discussions.

**Pull request merged**